rfcs/0003-currency.md

```diff
@@ -0,0 +1,116 @@
+
+# [DRAFT]
+
+Currency API.
+
+ Asset id as if it was deserialized, not necessary exists.
+ We could check asset id during serde, but that:
+ - will make serde setup complicated (need to write and consistently apply static singletons
+   to all places with asset id)
+ - validate will involve at minimum in memory cache call (in worth case db call) during
+   extrinsic invocation
+ - will need to disable this during calls when it is really no need for validation (new
+   currency mapping)
+ - normal path will pay price (validate each time), in instead when fail pays only (like
+   trying to transfer non existing asset id)
+ - we cannot guarantee existence of asset as it may be removed during transaction (so we
+   should make removal exclusive case)

```

Comment:
We could always create a wrapper type: `Exists<T: QueryExists>` and if your API needs asset_ids to be valid, it could ask for a `Exists<CurrencyId>`, where `impl QueryExists for CurrencyId` would perform the queries needed. However if the inner function does actions such as transfer etc, it is a waste of storage calls.

Reply:
Yes. So all apis will work with may be exists as they do now. Just need to make it clear. in some cases maybe would be cast into AssetIdForSure to act on that. But i am not yeat aware of this case. So in most cases Iam about zero waste.  Assuming that eventually and module checks currncy exists and we do not check it on entry..Non checking at all is attack with spam and garbage. This way may have well known ids to avoid storage check.